[examples] add controlnet sd3 example

DavyMorgan opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #8834


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @.

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Core library:

- Schedulers: @yiyixuxu
- Pipelines and pipeline callbacks: @yiyixuxu and @asomoza
- Training examples: @sayakpaul
- Docs: @stevhliu and @sayakpaul
- JAX and MPS: @pcuenca
- Audio: @sanchit-gandhi
- General functionalities: @sayakpaul @yiyixuxu @DN6

Integrations:

- deepspeed: HF Trainer/Accelerate: @SunMarc
- PEFT: @sayakpaul @BenjaminBossan

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- transformers: [different repo](https://github.com/huggingface/transformers)
- safetensors: [different repo](https://github.com/huggingface/safetensors)

-->

DavyMorgan commented:
My implementation is based on the official examples for [ControlNet](https://github.com/huggingface/diffusers/blob/main/examples/controlnet/README_sdxl.md) and [SD3 DreamBooth](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/README_sd3.md)

kadirnar commented:
@DavyMorgan 

Great. Are you going to write controlnet train code for Flux? I really need it.

DavyMorgan commented:
> @DavyMorgan
> 
> Great. Are you going to write controlnet train code for Flux? I really need it.

Hi, a nice implementation of controlnet flux with training scripts can be found at https://github.com/XLabs-AI/x-flux.

kadirnar commented:
> > @DavyMorgan
> > Great. Are you going to write controlnet train code for Flux? I really need it.
> 
> Hi, a nice implementation of controlnet flux with training scripts can be found at https://github.com/XLabs-AI/x-flux.

I have been using this library for 1 week and multi gpu is not working. I need multi gpu support to train large datasets.

DavyMorgan commented:
@yiyixuxu Could you take a look at this PR when you get a chance? It addresses issue #8834 and provides an example for controlnet+sd3. Thanks!

xduzhangjiayu commented:
> My implementation is based on the official examples for [ControlNet](https://github.com/huggingface/diffusers/blob/main/examples/controlnet/README_sdxl.md) and [SD3 DreamBooth](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/README_sd3.md)

Hi,
I always got this error when using the script
`TypeError: Can only concatenate tensors but got <class 'str'>`
Because the `example["prompts"]` is a string, do you have any advice about this? thanks.

yiyixuxu commented:
ohh thanks for your PR!

@haofanwang @wangqixun would you be able to give this PR a review too?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_9249). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

sayakpaul commented on examples/controlnet/README_sd3.md:
```diff
@@ -0,0 +1,126 @@
+
+## Training
+
+First download the SD3 model from [Hugging Face Hub](https://huggingface.co/stabilityai/stable-diffusion-3-medium). We will use it as a base model for the ControlNet training.
```
Could we add a similar note like this? You can reuse from https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/README_sd3.md. 

<img width="1136" alt="image" src="https://github.com/user-attachments/assets/7368c7a0-8947-4b3e-b87c-2b3bf7389b5e">

DavyMorgan replied:
done

DavyMorgan replied:
done

sayakpaul commented on examples/controlnet/README_sd3.md:
````diff
@@ -0,0 +1,126 @@
+Then run `huggingface-cli login` to log into your Hugging Face account. This is needed to be able to push the trained ControlNet parameters to Hugging Face Hub.
+
+```bash
+export MODEL_DIR="path to sd3 model"
````
Let's just put `stabilityai/stable-diffusion-3-medium-diffusers` here and a better default value to `OUTPUT_DIR` (`sd3-controlnet-out`).

DavyMorgan replied:
done

sayakpaul commented on examples/controlnet/README_sd3.md:
```diff
@@ -0,0 +1,126 @@
+from diffusers.utils import load_image
+import torch
+
+base_model_path = "path to sd3 model"
+controlnet_path = "path to controlnet ckpt"
```
Same. We could use a demo ControlNet checkpoint here. Maybe you have one since you were running experiments?

DavyMorgan replied:
done

sayakpaul commented on examples/controlnet/README_sd3.md:
```diff
@@ -0,0 +1,126 @@
+
+### GPU usage
+
+SD3 is a large model and requires a lot of GPU memory. 
+We recommend using two GPUs with at least 80GB of memory each, as indicated in the training instructions above.
+Specifically, one GPU is used for the ControlNet-SD3 training, and the other is used for validation.
+In this example, we set `CUDA_VISIBLE_DEVICES="0,1"` to use the first two GPUs on the machine.
+Make sure to use GPU 0 when configuring the accelerator, since the validation will use GPU 1.
```
1. Let's add a link to `accelerate config`: https://huggingface.co/docs/transformers/en/accelerate
2. In a follow-up PR, we could try LoRAs on ControlNet, that would help cut the memory requirements quite a bit. WDYT?

DavyMorgan replied:
Two GPUs are no longer needed since text embeddings are computed before constructing the DataLoader now (just as the official example of ControlNet-sdxl). Thanks for your valuable suggestion. The script can run smoothly with one single GPU.

sayakpaul commented on examples/controlnet/train_controlnet_sd3.py:
```diff
@@ -0,0 +1,1396 @@
+        else:
+            logger.warning(f"image logging not implemented for {tracker.name}")
+
+        del pipeline
+        gc.collect()
+        torch.cuda.empty_cache()
```
Could use:
https://github.com/huggingface/diffusers/blob/8ecf499d8bda3721ce89f5cb8c804afec4966b6a/src/diffusers/training_utils.py#L263

DavyMorgan replied:
done

sayakpaul commented on examples/controlnet/train_controlnet_sd3.py:
```diff
@@ -0,0 +1,1396 @@
+        variant=args.variant,
+        torch_dtype=weight_dtype,
+    )
+    pipeline = pipeline.to(torch.device('cuda:1'))
```
Note that since we run `log_validation()` under `accelerator.is_main_process`, we should use `accelerator.device` instead here.

DavyMorgan replied:
done

sayakpaul commented on examples/controlnet/train_controlnet_sd3.py:
```diff
@@ -0,0 +1,1396 @@
+    if args.seed is None:
+        generator = None
+    else:
+        generator = torch.Generator(device=torch.device('cuda:1')).manual_seed(args.seed)
```
Let's just do `torch.manual_seed()`?

DavyMorgan replied:
done

sayakpaul commented on examples/controlnet/train_controlnet_sd3.py:
```diff
@@ -0,0 +1,1396 @@
+        )
+
+    image_logs = []
+    inference_ctx = contextlib.nullcontext() if is_final_validation else torch.autocast("cuda")
```
```suggestion
    inference_ctx = contextlib.nullcontext() if is_final_validation else torch.autocast(accelerator.device.type)
```

sayakpaul commented on examples/controlnet/train_controlnet_sd3.py:
```diff
@@ -0,0 +1,1396 @@
+    tokenizers = [tokenizer_one, tokenizer_two, tokenizer_three]
+    text_encoders = [text_encoder_one, text_encoder_two, text_encoder_three]
+
+    def compute_text_embeddings(prompt, text_encoders, tokenizers):
```
Would it be possible to update the script to use `compute_text_embeddings()` after the dataset is prepared? This way, we won't have to keep the text encoders in GPU memory during the course of training. 

Or, we could offload the text encoders to CPU (assuming enough RAM) after text embeddings are computed and then load them back as needed. 

WDYT?

DavyMorgan replied:
Yeah, you are right. I have updated the codes, and now we don't keep the text encoders in GPU memory.

xduzhangjiayu replied:
Hi, 
I have tried the script but I cannot remove text encoders out of GPU before training the controlnet, did you test this feature successfully? 
Thanks so much!

DavyMorgan replied:
Yes. In line 1133 of the code, text encoders and tokenizers have been deleted before training. https://github.com/huggingface/diffusers/blob/6dc6486565ea1d8d1be567eefc1094e9185560a1/examples/controlnet/train_controlnet_sd3.py#L1133C5-L1133C61

xduzhangjiayu replied:
It cannot work on my side, the text encoders still occupy ~10GB memory and I don't know why.
Another question is do you know where the text embedding (output of text encoders) saved ? In RAM or Disk?
Thanks so much for the reply!

DavyMorgan replied:
Are you sure it is the text encoders that occupy the memory? The GPU memory can be other models, as the vae, transformer, and controlnet models are still in memory. Besides, as we periodically run the validation, the text encoders will also be loaded every `validation_steps` steps. From my experiments, previously I need to separate the training and validation in two distinct GPUS, and after the above update I only need one GPU to run the script.

During training, the text embeddings from text encoders are in memory, though there is a cached one in disk such that it will not compute them in your next run as long as the configs are the same.

sayakpaul commented on examples/controlnet/train_controlnet_sd3.py:
```diff
@@ -0,0 +1,1396 @@
+        weight_dtype = torch.bfloat16
+
+    # Move vae, transformer and text_encoder to device and cast to weight_dtype
+    vae.to(accelerator.device, dtype=weight_dtype)
```
We usually keep the `vae` in float32:
https://github.com/huggingface/diffusers/blob/8ecf499d8bda3721ce89f5cb8c804afec4966b6a/examples/dreambooth/train_dreambooth_lora_sd3.py#L1187

To prevent numerical instabilities. Did you not experience it in your experiments?

DavyMorgan replied:
Thanks for pointing this out. I haven't met numerical instabilities in my experiments, but I have updated the codes to use float32 according to your comment.

sayakpaul reviewed (commented):
Thank you for working on this! I have left a couple of comments. 

Could you also share some results from your experiments? 

And as @yiyixuxu mentioned, it'd be great to have this PR reviewed by @haofanwang @wangqixun as they were the first ones to have come up with SD3 ControlNets.

xduzhangjiayu commented:
> My implementation is based on the official examples for [ControlNet](https://github.com/huggingface/diffusers/blob/main/examples/controlnet/README_sdxl.md) and [SD3 DreamBooth](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/README_sd3.md)

Hi,
I tried run your script with mixed-precision=fp16, but got unexpected error in 
`                model_pred = transformer(
                    hidden_states=noisy_model_input,
                    timestep=timesteps,
                    encoder_hidden_states=prompt_embeds,
                    pooled_projections=pooled_prompt_embeds,
                    block_controlnet_hidden_states=control_block_res_samples,
                    return_dict=False,
                )[0]`
the error is `RuntimeError: mat1 and mat2 must have the same dtype, but got Float and Half`
But the `noisy_model_input` is fp16 in the script, do you have any advice to solve this problem? thanks so much!

DavyMorgan commented on examples/controlnet/train_controlnet_sd3.py:
```diff
@@ -0,0 +1,1396 @@
+                    timestep=timesteps,
+                    encoder_hidden_states=prompt_embeds,
+                    pooled_projections=pooled_prompt_embeds,
+                    block_controlnet_hidden_states=control_block_res_samples,
```
done

DavyMorgan commented:
@sayakpaul @xduzhangjiayu Thank you very much for your kind reviews. I have updated the codes according to your comments and suggestions. I have also added two experimental images as results in README.

sayakpaul commented:
Thanks for the changes @DavyMorgan! Let's also add a test similar to https://github.com/huggingface/diffusers/blob/main/examples/controlnet/test_controlnet.py?

DavyMorgan commented:
> Thanks for the changes @DavyMorgan! Let's also add a test similar to https://github.com/huggingface/diffusers/blob/main/examples/controlnet/test_controlnet.py?

@sayakpaul Yeah sure! I have added a similar test to https://github.com/huggingface/diffusers/blob/main/examples/controlnet/test_controlnet.py.

DavyMorgan commented:
> Could you run `make style && make quality`?

@sayakpaul Yeah. I have run `make style && make quality` which fixes the import order and a few style issues. Thanks a lot!

DavyMorgan commented:
```
09/09/2024 13:20:09 - INFO - __main__ - Initializing controlnet weights from transformer
Traceback (most recent call last):
  File "/__w/diffusers/diffusers/examples/controlnet/train_controlnet_sd3.py", line 1415, in <module>
    main(args)
  File "/__w/diffusers/diffusers/examples/controlnet/train_controlnet_sd3.py", line 997, in main
    controlnet = SD3ControlNetModel.from_transformer(transformer)
  File "/__w/diffusers/diffusers/src/diffusers/models/controlnet_sd3.py", line 254, in from_transformer
    controlnet.transformer_blocks.load_state_dict(transformer.transformer_blocks.state_dict(), strict=False)
  File "/opt/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 2215, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for ModuleList:
	size mismatch for 0.norm1_context.linear.weight: copying a param with shape torch.Size([64, 32]) from checkpoint, the shape in current model is torch.Size([192, 32]).
	size mismatch for 0.norm1_context.linear.bias: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([192]).
```
@sayakpaul It seems that the `hf-internal-testing/tiny-sd3-pipe` has size mismatch with SD3ControlNetModel. Should I use "InstantX/SD3-Controlnet-Canny" in the test script, while I find all the test codes use pretrained models from hf-internal-testing?

sayakpaul commented:
We need to use a smaller ControlNet model. We should be able to initialize that from the `transformer` of `hf-internal-testing/tiny-sd3-pipe`, as followed in the https://github.com/huggingface/diffusers/blob/main/examples/controlnet/test_controlnet.py script.

DavyMorgan commented:
> We need to use a smaller ControlNet model. We should be able to initialize that from the `transformer` of `hf-internal-testing/tiny-sd3-pipe`, as followed in the https://github.com/huggingface/diffusers/blob/main/examples/controlnet/test_controlnet.py script.

@sayakpaul Thanks. I have updated the test to leverage the smaller SD3 model used in the official test script in https://github.com/huggingface/diffusers/blob/f28a8c257afe8eeb16b4deb973c6b1829f6aea59/tests/pipelines/controlnet_sd3/test_controlnet_sd3.py#L172.

DavyMorgan commented:
I see. I have also added a tiny controlnet model based on the official test script of controlnet-sd3. Now the example test passes on my local machine. @sayakpaul

sayakpaul reviewed (approved):
Thanks for your contributions!

DavyMorgan commented on examples/controlnet/README_sd3.md:
```diff
@@ -0,0 +1,152 @@
+import torch
+
+base_model_path = "stabilityai/stable-diffusion-3-medium-diffusers"
+controlnet_path = "sd3-controlnet-out/checkpoint-6500/controlnet"
```
Yeah sure, I will upload my checkpoint to the hub.

DavyMorgan commented on examples/controlnet/README_sd3.md:
```diff
@@ -0,0 +1,152 @@
+
+#### After 6500 steps with batch size 8:
+
+| |  |
+|-------------------|:-------------------------:|
+|| pale golden rod circle with old lace background |
+ ![conditioning image](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/controlnet_training/conditioning_image_1.png) | ![pale golden rod circle with old lace background](https://huggingface.co/datasets/DavyMorgan/sd3-controlnet-results/resolve/main/step-6500.png) |
```
I guess I can include more sample images for validation.

DavyMorgan commented on examples/controlnet/train_controlnet_sd3.py:
```diff
@@ -0,0 +1,1416 @@
+logger = get_logger(__name__)
+
+
+def image_grid(imgs, rows, cols):
+    assert len(imgs) == rows * cols
+
+    w, h = imgs[0].size
+    grid = Image.new("RGB", size=(cols * w, rows * h))
+
+    for i, img in enumerate(imgs):
+        grid.paste(img, box=(i % cols * w, i // cols * h))
+    return grid
```
Got it! Thanks.

xduzhangjiayu commented:
Yes，I'm sure it is text encoders that occupy the memory。After computing the embedding，text encoders are still in the GPU memory. （That means clear_objs_and_retain_memory can't work）.I think it may related to my accelerate config. I will check later. Have you tried training with a large dataset（at least 1000k image)？ I think it may needs lots of CPU RAM when pre-computing the text embedding.



---- Replied Message ----
| From | Yu ***@***.***> |
| Date | 09/13/2024 17:43 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [huggingface/diffusers] [examples] add controlnet sd3 example (PR #9249) |

@DavyMorgan commented on this pull request.

In examples/controlnet/train_controlnet_sd3.py:

> +    )
+
+    train_dataset = make_train_dataset(args, tokenizer_one, tokenizer_two, tokenizer_three, accelerator)
+
+    train_dataloader = torch.utils.data.DataLoader(
+        train_dataset,
+        shuffle=True,
+        collate_fn=collate_fn,
+        batch_size=args.train_batch_size,
+        num_workers=args.dataloader_num_workers,
+    )
+
+    tokenizers = [tokenizer_one, tokenizer_two, tokenizer_three]
+    text_encoders = [text_encoder_one, text_encoder_two, text_encoder_three]
+
+    def compute_text_embeddings(prompt, text_encoders, tokenizers):


Are you sure it is the text encoders that occupy the memory? The GPU memory can be other models, as the vae, transformer, and controlnet models are still in memory. Besides, as we periodically run the validation, the text encoders will also be loaded every validation_steps steps. From my experiments, previously I need to separate the training and validation in two distinct GPUS, and after the above update I only need one GPU to run the script.

During training, the text embeddings from text encoders are in memory, though there is a cached one in disk such that it will not compute them in your next run as long as the configs are the same.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

DavyMorgan commented:
> Yes，I'm sure it is text encoders that occupy the memory。After computing the embedding，text encoders are still in the GPU memory. （That means clear_objs_and_retain_memory can't work）.I think it may related to my accelerate config. I will check later. Have you tried training with a large dataset（at least 1000k image)？ I think it may needs lots of CPU RAM when pre-computing the text embedding.
> ---- Replied Message ---- | From | Yu ***@***.***> | | Date | 09/13/2024 17:43 | | To | ***@***.***> | | Cc | ***@***.***>***@***.***> | | Subject | Re: [huggingface/diffusers] [examples] add controlnet sd3 example (PR #9249) | @DavyMorgan commented on this pull request. In examples/controlnet/train_controlnet_sd3.py:
> + )
> + + train_dataset = make_train_dataset(args, tokenizer_one, tokenizer_two, tokenizer_three, accelerator) + + train_dataloader = torch.utils.data.DataLoader( + train_dataset, + shuffle=True, + collate_fn=collate_fn, + batch_size=args.train_batch_size, + num_workers=args.dataloader_num_workers, + ) + + tokenizers = [tokenizer_one, tokenizer_two, tokenizer_three] + text_encoders = [text_encoder_one, text_encoder_two, text_encoder_three] + + def compute_text_embeddings(prompt, text_encoders, tokenizers): Are you sure it is the text encoders that occupy the memory? The GPU memory can be other models, as the vae, transformer, and controlnet models are still in memory. Besides, as we periodically run the validation, the text encoders will also be loaded every validation_steps steps. From my experiments, previously I need to separate the training and validation in two distinct GPUS, and after the above update I only need one GPU to run the script. During training, the text embeddings from text encoders are in memory, though there is a cached one in disk such that it will not compute them in your next run as long as the configs are the same. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were mentioned.Message ID: ***@***.***>

I only tested the fill50k dataset. As we use the `datasets` library, I believe it will handle the memory/disk issue well. You can check https://huggingface.co/docs/datasets/en/about_arrow#memory-mapping.

xduzhangjiayu commented:
> > Yes，I'm sure it is text encoders that occupy the memory。After computing the embedding，text encoders are still in the GPU memory. （That means clear_objs_and_retain_memory can't work）.I think it may related to my accelerate config. I will check later. Have you tried training with a large dataset（at least 1000k image)？ I think it may needs lots of CPU RAM when pre-computing the text embedding.
> > ---- Replied Message ---- | From | Yu _**@**_._**> | | Date | 09/13/2024 17:43 | | To | _**@**_.**_> | | Cc | _**@**_._**>**_@_**.**_> | | Subject | Re: [huggingface/diffusers] [examples] add controlnet sd3 example (PR #9249) | @DavyMorgan commented on this pull request. In examples/controlnet/train_controlnet_sd3.py:
> > 
> > * )
> > * * train_dataset = make_train_dataset(args, tokenizer_one, tokenizer_two, tokenizer_three, accelerator) + + train_dataloader = torch.utils.data.DataLoader( + train_dataset, + shuffle=True, + collate_fn=collate_fn, + batch_size=args.train_batch_size, + num_workers=args.dataloader_num_workers, + ) + + tokenizers = [tokenizer_one, tokenizer_two, tokenizer_three] + text_encoders = [text_encoder_one, text_encoder_two, text_encoder_three] + + def compute_text_embeddings(prompt, text_encoders, tokenizers): Are you sure it is the text encoders that occupy the memory? The GPU memory can be other models, as the vae, transformer, and controlnet models are still in memory. Besides, as we periodically run the validation, the text encoders will also be loaded every validation_steps steps. From my experiments, previously I need to separate the training and validation in two distinct GPUS, and after the above update I only need one GPU to run the script. During training, the text embeddings from text encoders are in memory, though there is a cached one in disk such that it will not compute them in your next run as long as the configs are the same. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were mentioned.Message ID: _**@**_.***>
> 
> I only tested the fill50k dataset. As we use the `datasets` library, I believe it will handle the memory/disk issue well. You can check https://huggingface.co/docs/datasets/en/about_arrow#memory-mapping.

OK, once again thank you very much for your reply !
